conda_build/metadata.py

```diff
@@ -907,6 +953,63 @@ def toposort(output_metadata_map):
     return result
 
 
+def _toposort_outputs(output_tuples: list[OutputTuple]) -> list[OutputTuple]:
+    """This function is used to work out the order to run the install scripts
+    for split packages based on any interdependencies. The result is just
+    a re-ordering of outputs such that we can run them in that order and
+    reset the initial set of files in the install prefix after each. This
+    will naturally lead to non-overlapping files in each package and also
+    the correct files being present during the install and test procedures,
+    provided they are run in this order."""
+    from conda.common.toposort import _toposort
+
+    # We only care about the conda packages built by this recipe. Non-conda
+    # packages get sorted to the end.
+    conda_outputs: dict[str, list[OutputTuple]] = {}
+    non_conda_outputs: list[OutputTuple] = []
+    for output_tuple in output_tuples:
+        output_d, _ = output_tuple
+        if output_d.get("type", "conda").startswith("conda"):
+            # the same package name may be seen multiple times (variants)
+            conda_outputs.setdefault(output_d["name"], []).append(output_tuple)
+        else:
+            non_conda_outputs.append(output_tuple)
+
+    # Iterate over conda packages, creating a mapping of package names to their
+    # dependencies to be used in toposort
+    name_to_dependencies: dict[str, set[str]] = {}
+    name_lookup: dict[str, list[OutputTuple]] = {}
+    for name, same_name_outputs in conda_outputs.items():
+        for output_d, output_metadata in same_name_outputs:
+            # dependencies for all of the variants
+            dependencies = (
+                *output_metadata.get_value("requirements/run", []),
+                *output_metadata.get_value("requirements/host", []),
+                *(
+                    output_metadata.get_value("requirements/build", [])
+                    if not output_metadata.is_cross
+                    else []
+                ),
+            )
+            name_to_dependencies.setdefault(name, set()).update(
+                dependency_name
+                for dependency in dependencies
+                if (dependency_name := dependency.split(" ")[0]) in conda_outputs
+            )
+
+            # preserve all of the variants
+            name_lookup.setdefault(name, []).append((output_d, output_metadata))
+
+    return [
+        *(
+            output
+            for name in _toposort(name_to_dependencies)
+            for output in name_lookup[name]
+        ),
+        *non_conda_outputs,
```

Comment:
I like the new code better than the old, for sure.
But I'm confused by the old code/comment
```
    # not sure that this is working...  not everything has 'name', and not sure how this pans out
    #    may end up excluding packages without the 'name' field
    ...
```
which indeed left out things `without the 'name' field`.
I didn't much read the surrounding code so have no idea when/how such cases would occur (if they could occur?) and also how the code after the toposort call would then handle/not handle this.
I'm confused :confused: ...

Reply:
Hm should we return to the old way of ignoring outputs without names but be more explicit?

```python
for output_tuple in output_tuples:
    output_d, _ = output_tuple
    if output_d.get("type", "conda").startswith("conda"):
        # the same package name may be seen multiple times (variants)
        conda_outputs.setdefault(output_d["name"], []).append(output_tuple)
    elif "name" in output_d:
        non_conda_outputs.append(output_tuple)
    else:
        log.warn("Found a non-conda package output without a name, ignoring")
```

Reply:
Since this is a regression fix I will switch to the above to avoid any other side-effects of this fix

Reply:
Done https://github.com/conda/conda-build/pull/5345/commits/6e0119944bc0e93520c9bfaa5a19bda1e57103ea

Reply:
Thanks! The change, including the very valid `is it even possible to get here?` comment is perfect.